src/dmd/semantic3.d

```diff
@@ -1356,3 +1346,36 @@ private extern(C++) final class Semantic3Visitor : Visitor
         ad.semanticRun = PASS.semantic3done;
     }
 }
+
+private struct FuncDeclSem3
+{
+    // The FuncDeclaration subject to Semantic analysis
+    FuncDeclaration funcdecl;
+
+    // Scope of analysis
+    Scope* sc;
+    this(FuncDeclaration fd,Scope* s)
+    {
+        funcdecl = fd;
+        sc = s;
+    }
+
+    /* Checks that the overriden functions (if any) have in contracts if
+     * funcdecl has an in contract.
+     */
+    void checkInContractOverrides()
+    {
+        if (funcdecl.frequires)
+        {
+            for (size_t i = 0; i < funcdecl.foverrides.dim; i++)
+            {
+                FuncDeclaration fdv = funcdecl.foverrides[i];
+                if (fdv.fbody && !fdv.frequires)
+                {
+                    funcdecl.error("cannot have an in contract when overridden function `%s` does not have an in contract", fdv.toPrettyChars());
+                    break;
+                }
+            }
+        }
+    }
+}
```

Comment:
I see the point of factoring out checkInContractOverrides() into a separate function, but I don't see the point of creating FuncDeclSem3?

Reply:
It will hold local variables of `visit(FuncDeclaration)` so that there won't be heaps of arguments to the function.

Reply:
We already have `Scope` for that.